samcli/lib/config/samconfig.py

```diff
@@ -148,24 +151,38 @@ def config_dir(template_file_path=None):
         return os.getcwd()
 
     def _read(self):
-        if self.document:
-            return self.document
-
-        try:
-            txt = self.filepath.read_text()
-            self.document = tomlkit.loads(txt)
-        except OSError:
-            self.document = tomlkit.document()
-
+        if not self.document:
+            try:
+                txt = self.filepath.read_text()
+                self.document = tomlkit.loads(txt)
+                self._version_sanity_check(self._version())
+            except OSError:
+                self.document = tomlkit.document()
+
+        if self.document.body:
+            self._version_sanity_check(self._version())
         return self.document
 
     def _write(self):
         if not self.document:
             return
         if not self.exists():
             open(self.filepath, "a+").close()
+        current_version = self._version() if self._version() else SAM_CONFIG_VERSION
+        try:
+            self.document.add(VERSION_KEY, current_version)
+        except tomlkit.exceptions.KeyAlreadyPresent:
+            # NOTE(TheSriram): Do not attempt to re-write an existing version
+            pass
         self.filepath.write_text(tomlkit.dumps(self.document))
 
+    def _version(self):
+        return self.document.get(VERSION_KEY, None)
+
+    def _version_sanity_check(self, version):
+        if not isinstance(version, float):
+            raise SamConfigVersionException(f"'{VERSION_KEY}' key is not present " f"or is in unrecognized format. ")
```

Comment:
This is well and good, but what is the purpose of a version number? The AWS CLI config has never needed this, what does it mean when we increment it?

Reply:
This was discussed as part of design, but at the very least, it provides us a mechanism depending on the growth of scope of the configuration file in the future.